syntax = "proto3";

package diablopb;

option java_package = "com.diablo.grpc";
option java_outer_classname = "InventoryProto";
option java_multiple_files = true;

service InventoryService {
    // 1. UNARY
    rpc GetItem (ItemRequest) returns (ItemResponse);
    // 2. SERVER STREAMING
    rpc ListItemsByType (ItemTypeRequest) returns (stream ItemResponse);
    // 3. CLIENT STREAMING
    rpc CreateBulkLoot (stream CreateItemRequest) returns (BulkCreateResponse);
}

message ItemRequest {
    string id = 1;
}

message ItemTypeRequest {
    ItemType tipo = 1;
}

message CreateItemRequest {
    string id = 1;
    string nombre = 2;
    ItemType tipo = 3;
    int32 poder_de_objeto = 4;
    
    oneof datos_propios {
        int32 dano_base = 5;
        int32 armadura_base = 6;
    }

    string habilidad_pasiva = 7;
    string habilidad_activa = 8;
    repeated GemType gemas = 9;
}

message ItemResponse {
    string id = 1;
    string nombre = 2;
    ItemType tipo = 3;
    int32 poder_de_objeto = 4;
    
    oneof datos_propios {
        int32 dano_base = 5;
        int32 armadura_base = 6;
    }

    string habilidad_pasiva = 7;
    string habilidad_activa = 8;
    repeated GemType gemas_incrustadas = 9;
}

message BulkCreateResponse {
    int32 items_creados = 1;
    string mensaje = 2;
    repeated ItemResponse items = 3;
}

enum ItemType {
    UNKNOWN_TYPE = 0;
    WEAPON = 1;
    ARMOR = 2;
}

enum GemType {
    UNKNOWN_GEM = 0;
    AMATISTA = 1;
    CRANEO = 2;
    DIAMANTE = 3;
    ESMERALDA = 4;
    RUBI = 5;
    TOPACIO = 6;
    ZAFIRO = 7;
}
